File: edge/src/lab/publish-requests.ts
import {
  body,
  candidates,
  fail,
  json,
  objectResponse,
  type ReadyEnv,
  type Row,
} from './common.js'
import { normalizedHolds, publishCloudbase } from './cloudbase.js'
export function requestView(r: Row) {
  return {
    id: r.id,
    applicantId: r.applicant_id,
    wallName: r.wall_name,
    target: r.target,
    status: r.status,
    createdAt: r.created_at / 1000,
    reason: r.reason ?? '',
    error: r.error ?? '',
    retryable:
      r.status === 'failed' ||
      (r.status === 'publishing' && Number(r.lease_until) <= Date.now()),
    reviewerId: r.reviewer_id ?? null,
    reviewedAt: r.reviewed_at ? r.reviewed_at / 1000 : null,
    result: r.result ? JSON.parse(r.result) : null,
  }
}
export async function createPublishRequest(
  env: ReadyEnv,
  e: Row,
  c: Row,
  user: Row,
  b: Row,
) {
  if (b.target !== 'cloudbase') fail('INVALID_TARGET', '该目标不支持发布申请。')
  const name = String(b.wallName ?? `${e.name} · 校准`).trim()
  if (!name || name.length > 120)
    fail('INVALID_INPUT', '墙面名称须为 1–120 字。')
  const prior = await env.DB.prepare(
    "SELECT * FROM lab_publish_requests WHERE calibration_id=? AND target=? AND status<>'rejected'",
  )
    .bind(c.id, b.target)
    .first<Row>()
  if (prior) return prior
  const source = await env.MEDIA.get(c.candidates_key),
    image = await env.MEDIA.get(c.display_key)
  if (!source || !image) fail('NOT_FOUND', '校准文件不存在。', 404)
  const holds = candidates(
    JSON.parse(await source.text()).items,
    e.width,
    e.height,
  )
  if (!holds.length) fail('INVALID_CANDIDATES', '请至少保存一个岩点。')
  // Preflight the receiver contract before retaining a non-editable review snapshot.
  normalizedHolds(holds, e.width, e.height)
  const id = crypto.randomUUID(),
    key = `lab-publish-requests/${id}/`
  try {
    await env.MEDIA.put(
      key + 'snapshot.json',
      JSON.stringify({
        publishRequestId: `cloudflare:${id}`,
        sourceExperimentId: e.id,
        sourceCalibrationId: c.id,
        wallName: name,
        imageWidth: e.width,
        imageHeight: e.height,
        holds,
      }),
      { httpMetadata: { contentType: 'application/json' } },
    )
    await env.MEDIA.put(key + 'display.webp', image.body, {
      httpMetadata: { contentType: 'image/webp' },
    })
    const row = await env.DB.prepare(
      "INSERT OR IGNORE INTO lab_publish_requests (id,applicant_id,experiment_id,calibration_id,wall_name,target,status,snapshot_key,created_at) VALUES (?,?,?,?,?,?,'pending',?,?) RETURNING *",
    )
      .bind(id, user.id, e.id, c.id, name, b.target, key, Date.now())
      .first<Row>()
    if (row) return row
    await env.MEDIA.delete([key + 'snapshot.json', key + 'display.webp'])
    return (await env.DB.prepare(
      "SELECT * FROM lab_publish_requests WHERE calibration_id=? AND target=? AND status<>'rejected'",
    )
      .bind(c.id, b.target)
      .first<Row>())!
  } catch (error) {
    await env.MEDIA.delete([key + 'snapshot.json', key + 'display.webp'])
    throw error
  }
}
export async function approvePublishRequest(
  env: ReadyEnv,
  id: string,
  reviewerId: string,
) {
  const token = crypto.randomUUID(),
    now = Date.now()
  const row = await env.DB.prepare(
    "UPDATE lab_publish_requests SET status='publishing',lease_token=?,lease_until=?,error=NULL,reviewer_id=?,reviewed_at=? WHERE id=? AND (status IN ('pending','failed') OR (status='publishing' AND lease_until<=?)) RETURNING *",
  )
    .bind(token, now + 600000, reviewerId, now, id, now)
    .first<Row>()
  if (!row) {
    const existing = await env.DB.prepare(
      'SELECT * FROM lab_publish_requests WHERE id=?',
    )
      .bind(id)
      .first<Row>()
    if (existing?.status === 'published') return existing
    fail('CONFLICT', '该申请正在处理或已被拒绝。', 409)
  }
  try {
    const snapshot = await env.MEDIA.get(row.snapshot_key + 'snapshot.json'),
      image = await env.MEDIA.get(row.snapshot_key + 'display.webp')
    if (!snapshot || !image) throw new Error('申请快照不存在。')
    const result = await publishCloudbase(
      env,
      JSON.parse(await snapshot.text()),
      await new Response(image.body).arrayBuffer(),
    )
    await env.DB.prepare(
      "UPDATE lab_publish_requests SET status='published',result=?,lease_token=NULL,lease_until=NULL WHERE id=? AND lease_token=?",
    )
      .bind(JSON.stringify(result), id, token)
      .run()
  } catch (error) {
    await env.DB.prepare(
      "UPDATE lab_publish_requests SET status='failed',error=?,lease_token=NULL,lease_until=NULL WHERE id=? AND lease_token=?",
    )
      .bind(
        error instanceof Error && ['LabError', 'Error'].includes(error.name)
          ? error.message.slice(0, 300)
          : 'CloudBase 发布失败，请管理员重试。',
        id,
        token,
      )
      .run()
  }
  return (await env.DB.prepare('SELECT * FROM lab_publish_requests WHERE id=?')
    .bind(id)
    .first<Row>())!
}
export async function publishRequestRoutes(
  request: Request,
  env: ReadyEnv,
  user: Row,
  path: string,
): Promise<Response | null> {
  const admin = user.role === 'admin'
  if (path === '/publish-requests' && request.method === 'GET') {
    const rows = await (
      admin
        ? env.DB.prepare(
            'SELECT * FROM lab_publish_requests ORDER BY created_at DESC',
          )
        : env.DB.prepare(
            'SELECT * FROM lab_publish_requests WHERE applicant_id=? ORDER BY created_at DESC',
          ).bind(user.id)
    ).all<Row>()
    return json({ items: rows.results.map(requestView), isAdmin: admin })
  }
  const m = path.match(
    /^\/publish-requests\/([\w-]+)\/(preview|image|approve|reject)$/,
  )
  if (!m) return null
  const r = await env.DB.prepare(
    'SELECT * FROM lab_publish_requests WHERE id=?',
  )
    .bind(m[1])
    .first<Row>()
  if (!r || (!admin && r.applicant_id !== user.id))
    fail('NOT_FOUND', '发布申请不存在。', 404)
  if (request.method === 'GET' && m[2] === 'image')
    return objectResponse(env, r.snapshot_key + 'display.webp', 'image/webp')
  if (request.method === 'GET' && m[2] === 'preview') {
    const object = await env.MEDIA.get(r.snapshot_key + 'snapshot.json')
    if (!object) fail('NOT_FOUND', '申请快照不存在。', 404)
    const s = JSON.parse(await object.text()),
      items = candidates(s.holds, s.imageWidth, s.imageHeight)
    const escaped = (v: string) =>
      v.replace(
        /[&<>"']/g,
        (c) =>
          ({
            '&': '&amp;',
            '<': '&lt;',
            '>': '&gt;',
            '"': '&quot;',
            "'": '&#39;',
          })[c]!,
      )
    return new Response(
      `<!doctype html><html lang="zh"><meta charset="utf-8"><title>${escaped(s.wallName)}</title><style>body{margin:0;background:#111;color:#fff;font:16px sans-serif}h1{font-size:18px;padding:12px}svg{display:block;width:100%;height:auto}polygon{fill:#77c94b44;stroke:#77c94b;stroke-width:2;vector-effect:non-scaling-stroke}</style><h1>${escaped(s.wallName)}</h1><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${s.imageWidth} ${s.imageHeight}"><image href="image" width="${s.imageWidth}" height="${s.imageHeight}" preserveAspectRatio="none"/>${items.map((h) => `<polygon points="${h.polygon.map((p: number[]) => p.join(',')).join(' ')}"/>`).join('')}</svg></html>`,
      {
        headers: {
          'Content-Type': 'text/html; charset=utf-8',
          'Cache-Control': 'no-store',
          'Content-Security-Policy':
            "default-src 'none'; img-src 'self'; style-src 'unsafe-inline'; frame-ancestors 'self'",
        },
      },
    )
  }
  if (request.method === 'POST' && (m[2] === 'approve' || m[2] === 'reject')) {
    if (!admin) fail('FORBIDDEN', '需要管理员权限。', 403)
    if (m[2] === 'approve')
      return json(requestView(await approvePublishRequest(env, r.id, user.id)))
    const b = await body(request),
      reason = String(b.reason ?? '').trim()
    if (reason.length > 300) fail('INVALID_INPUT', '拒绝原因不能超过 300 字。')
    const updated = await env.DB.prepare(
      "UPDATE lab_publish_requests SET status='rejected',reason=?,reviewer_id=?,reviewed_at=? WHERE id=? AND status='pending' RETURNING *",
    )
      .bind(reason, user.id, Date.now(), r.id)
      .first<Row>()
    if (!updated) fail('CONFLICT', '该申请正在处理或已结束。', 409)
    return json(requestView(updated))
  }
  return null
}

File: tools/segmentation-lab/src/segmentation_lab/publish_requests.py
"""Persistent approval snapshots, independent of the editable experiment tree."""
from contextlib import contextmanager
from html import escape
import fcntl
import json
import os
from pathlib import Path
import time
from uuid import uuid4

from fastapi import Body, HTTPException, Request
from fastapi.responses import FileResponse, HTMLResponse, Response

from .cloudbase_sync import build_normalized_holds
from .ownership import IDENTIFIER


def bounded_text(value, limit, field):
    if not isinstance(value, str) or not value.strip() or len(value.strip()) > limit or any(ord(c) < 32 for c in value):
        raise HTTPException(422, f'{field} must contain 1–{limit} characters without control characters.')
    return value.strip()


class PublishRequests:
    def __init__(self, root: Path):
        self.root = root
        root.mkdir(parents=True, exist_ok=True)

    @contextmanager
    def lock(self, name):
        # Kernel locks survive concurrent workers and are released after a crash.
        with (self.root / (name + '.lock')).open('a') as handle:
            try:
                fcntl.flock(handle, fcntl.LOCK_EX | fcntl.LOCK_NB)
            except BlockingIOError:
                raise HTTPException(409, 'This request is already being processed.')
            try:
                yield
            finally:
                fcntl.flock(handle, fcntl.LOCK_UN)

    def read(self, request_id):
        if not IDENTIFIER.fullmatch(request_id) or (self.root / request_id).is_symlink():
            raise HTTPException(404, 'Publish request not found.')
        try:
            return json.loads((self.root / request_id / 'request.json').read_text())
        except FileNotFoundError:
            raise HTTPException(404, 'Publish request not found.')

    def write(self, item):
        directory = self.root / item['id']
        temporary = directory / ('request-' + str(uuid4()) + '.tmp')
        with temporary.open('w') as handle:
            json.dump(item, handle, ensure_ascii=False, allow_nan=False)
            handle.flush()
            os.fsync(handle.fileno())
        temporary.replace(directory / 'request.json')

    def items(self):
        return sorted([json.loads(path.read_text()) for path in self.root.glob('*/request.json')], key=lambda item: item['createdAt'], reverse=True)


def install_publish_requests(app, store, settings, publish):
    requests = PublishRequests(settings.data_dir / 'publish-requests')

    def public(item):
        result = {key: value for key, value in item.items() if key not in {'metadata', 'imageName'}}
        result['retryable'] = item['status'] == 'failed'
        if item['status'] == 'publishing':
            try:
                with requests.lock(item['id']):
                    result['retryable'] = True
            except HTTPException as error:
                if error.status_code != 409:
                    raise
                result['retryable'] = False
        return result

    def accessible(request, request_id):
        item = requests.read(request_id)
        actor = request.state.actor
        if item.get('deletedAt') is not None or (not actor['isAdmin'] and item['applicantId'] != actor['userId']):
            raise HTTPException(404, 'Publish request not found.')
        return item

    def admin_only(request):
        if not request.state.actor['isAdmin']:
            raise HTTPException(403, 'Administrator access required.')

    @app.get('/api/publish-requests')
    def list_requests(request: Request):
        actor = request.state.actor
        return {'items': [public(item) for item in requests.items() if item.get('deletedAt') is None and (actor['isAdmin'] or item['applicantId'] == actor['userId'])], 'isAdmin': actor['isAdmin']}

    @app.post('/api/experiments/{experiment_id}/calibrations/{calibration_id}/publish-requests', status_code=201)
    def create_request(experiment_id: str, calibration_id: str, request: Request, payload: dict = Body(...)):
        target = payload.get('target')
        if not isinstance(target, str) or target not in {'cloudbase', 'cloudflare'}:
            raise HTTPException(422, 'Requests support cloudbase and cloudflare targets only.')
        name = bounded_text(payload.get('wallName'), 120, 'Wall name')
        with requests.lock('create'):
            for item in requests.items():
                if (item['applicantId'], item['sourceExperimentId'], item['sourceCalibrationId'], item['target']) == (request.state.actor['userId'], experiment_id, calibration_id, target) and item['status'] != 'rejected':
                    return public(item)
            experiment = next((item for item in store.list_experiments() if item['id'] == experiment_id), None)
            if experiment is None or not any(item['id'] == calibration_id for item in store.list_calibrations(experiment_id)):
                raise HTTPException(404, 'Calibration not found.')
            image_path = next((store.root / experiment_id / 'input').glob('original.*'), None)
            if image_path is None:
                raise HTTPException(404, 'Image not found.')
            holds = [{'sourceId': str(item['id']), 'kind': item.get('kind', 'hold'), 'polygon': item['polygon']} for item in store.read_calibration_candidates(experiment_id, calibration_id)]
            build_normalized_holds(holds, experiment['width'], experiment['height'])
            request_id = str(uuid4())
            metadata = {'publishRequestId': request_id, 'sourceExperimentId': experiment_id, 'sourceCalibrationId': calibration_id, 'wallName': name, 'imageWidth': experiment['width'], 'imageHeight': experiment['height'], 'holds': holds}
            item = {'id': request_id, 'applicantId': request.state.actor['userId'], 'sourceExperimentId': experiment_id, 'sourceCalibrationId': calibration_id, 'wallName': name, 'target': target, 'status': 'pending', 'createdAt': time.time(), 'reason': '', 'error': '', 'result': None, 'metadata': metadata, 'imageName': image_path.name}
            directory = requests.root / request_id
            directory.mkdir()
            with (directory / 'image').open('wb') as handle:
                handle.write(image_path.read_bytes())
                handle.flush()
                os.fsync(handle.fileno())
            requests.write(item)
            return public(item)

    @app.delete('/api/publish-requests/{request_id}', status_code=204)
    def delete_request(request_id: str, request: Request):
        accessible(request, request_id)
        with requests.lock(request_id):
            item = accessible(request, request_id)
            if item['status'] not in {'published', 'rejected'}:
                raise HTTPException(409, 'Only published or rejected requests can be deleted.')
            # Keep the receipt to prevent submitting the same calibration twice.
            item['deletedAt'] = time.time()
            requests.write(item)
        return Response(status_code=204)

    @app.get('/api/publish-requests/{request_id}/image')
    def image(request_id: str, request: Request):
        item = accessible(request, request_id)
        return FileResponse(requests.root / request_id / 'image', media_type='image/png' if item['imageName'].endswith('.png') else 'image/jpeg')

    @app.get('/api/publish-requests/{request_id}/preview')
    def preview(request_id: str, request: Request):
        item = accessible(request, request_id)
        metadata = item['metadata']
        width, height = metadata['imageWidth'], metadata['imageHeight']
        polygons = ''.join('<polygon points="' + ' '.join(f'{float(x)},{float(y)}' for x, y in hold['polygon']) + '"/>' for hold in metadata['holds'])
        return HTMLResponse(f'<!doctype html><html><meta charset="utf-8"><title>{escape(item["wallName"])}</title><style>body{{margin:24px;background:#111;color:#eee;font-family:sans-serif}}svg{{max-width:100%;height:auto}}polygon{{fill:#77c94b44;stroke:#77c94b;stroke-width:2;vector-effect:non-scaling-stroke}}</style><h1>{escape(item["wallName"])}</h1><svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}"><image href="image" width="{width}" height="{height}"/>{polygons}</svg></html>')

    @app.post('/api/publish-requests/{request_id}/approve')
    async def approve(request_id: str, request: Request):
        admin_only(request)
        accessible(request, request_id)
        with requests.lock(request_id):
            item = accessible(request, request_id)
            if item['status'] == 'published':
                return public(item)
            if item['status'] == 'rejected':
                raise HTTPException(409, 'Rejected requests cannot be approved.')
            # A publishing record with no active kernel lock is a crashed attempt.
            # Reuse its immutable metadata/id so destination deduplication recovers it.
            item.update(status='publishing', error='', reviewerId=request.state.actor['userId'], reviewedAt=time.time())
            requests.write(item)
            try:
                result = await publish(item['target'], (requests.root / request_id / 'image').read_bytes(), item['imageName'], item['metadata'])
                item.update(status='published', result=result)
            except Exception as error:
                item.update(status='failed', error=str(error)[:500])
            requests.write(item)
            return public(item)

    @app.post('/api/publish-requests/{request_id}/reject')
    def reject(request_id: str, request: Request, payload: dict = Body(...)):
        admin_only(request)
        accessible(request, request_id)
        reason = payload.get('reason', '')
        if not isinstance(reason, str) or len(reason.strip()) > 300 or any(ord(c) < 32 for c in reason):
            raise HTTPException(422, 'Reason must be at most 300 characters without control characters.')
        reason = reason.strip()
        with requests.lock(request_id):
            item = accessible(request, request_id)
            if item['status'] in {'publishing', 'published', 'failed'}:
                # A failed remote attempt can have committed before a timeout.
                raise HTTPException(409, 'A started publication must be recovered through approval.')
            item.update(status='rejected', reason=reason, reviewerId=request.state.actor['userId'], reviewedAt=time.time())
            requests.write(item)
            return public(item)
